pkg/cloudproviders/k8s/k8s.go

```diff
@@ -0,0 +1,375 @@
+}
+
+// NewProviderFromViper returns a new k8s provider.
+func NewProviderFromViper(v *viper.Viper, logger logrus.FieldLogger, version string) (gostatsd.CloudProvider, error) {
+	k := util.GetSubViper(v, "k8s")
+	setViperDefaults(k, version)
+
+	// Set up the k8s client
+	clientset, err := createKubernetesClient(k.GetString(ParamUserAgent), k.GetString(ParamKubeconfigPath),
+		k.GetString(ParamKubeconfigContext), k.GetFloat64(ParamAPIQPS), k.GetFloat64(ParamAPIQPSBurst))
+	if err != nil {
+		return nil, err
+	}
+
+	return NewProvider(
+		logger,
+		clientset,
+		PodInformerOptions{
+			ResyncPeriod: k.GetDuration(ParamResyncPeriod),
+			WatchCluster: k.GetBool(ParamWatchCluster),
+			NodeName:     k.GetString(ParamNodeName),
+		},
+		regexp.MustCompile(k.GetString(ParamAnnotationTagRegex)),
+		regexp.MustCompile(k.GetString(ParamLabelTagRegex)))
+}
+
+// NewProvider returns a new k8s provider.
+func NewProvider(logger logrus.FieldLogger, clientset kubernetes.Interface, podInfOpts PodInformerOptions,
+	annotationRegex, labelRegex *regexp.Regexp) (gostatsd.CloudProvider, error) {
+
+	// This list operation is for debugging purposes and failing fast. If this fails then the cache will likely fail.
+	_, err := clientset.CoreV1().Pods(meta_v1.NamespaceAll).List(meta_v1.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	// If we're not watching the entire cluster we need to limit our watch to pods with our node name
+	customWatchOptions := func(*meta_v1.ListOptions) {}
+	if !podInfOpts.WatchCluster {
+		if podInfOpts.NodeName == "" {
+			return nil, fmt.Errorf("watch-cluster set to false, and node name not supplied")
+		}
+		fieldSelector := "spec.nodeName=" + podInfOpts.NodeName
+		logger.WithField("fieldSelector", fieldSelector).Debug("set fieldSelector for informers")
+		customWatchOptions = func(lo *meta_v1.ListOptions) {
+			lo.FieldSelector = fieldSelector
+		}
+	}
+
+	// Create a shared informer factory that watches the correct nodes
+	factory := informers.NewSharedInformerFactoryWithOptions(
+		clientset,
+		podInfOpts.ResyncPeriod,
+		informers.WithTweakListOptions(customWatchOptions))
+
+	// Set up the pod informer which fills an index with the pods we care about
+	podsInf := factory.Core().V1().Pods().Informer()
+	indexers := cache.Indexers{
+		PodsByIPIndexName: podByIpIndexFunc,
+	}
+	err = podsInf.AddIndexers(indexers)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO: we should emit prometheus metrics to fit in with the k8s ecosystem
+	// TODO: we should emit events to the k8s API to fit in with the k8s ecosystem
+
+	return &Provider{
+		logger:          logger,
+		podsInf:         podsInf,
+		factory:         factory,
+		annotationRegex: annotationRegex,
+		labelRegex:      labelRegex,
+	}, nil
+}
+
+func (p *Provider) Run(ctx context.Context) {
+	go p.factory.Start(ctx.Done()) // fork
+	// wait for the initial synchronization of the local cache.
+	// TODO: why is the error message firing for our tests?
```

Comment:
Perhaps you don't have the watch part mocked properly? IIRC some extra plumbing may be necessary for it to work. Smith probably has an example of this working.

Reply:
I actually fixed this I just forgot to remove the TODO. My bad.